### DISCUSSION TITLE
WSL usbipd passthrough - No Lime devices found

### DISCUSSION THREAD

**Discussion by Reshims** [LELEC210X/LELEC210X#34] [Q&A] [answered]
# Problem

Hello,

I've been using WSL 2.0 with Ubuntu 20.04 for the project. 
Following the tutorial here on github ([INSTALL.md](https://github.com/LELEC210X/LELEC210X/blob/main/INSTALL.md)), everything works fine, but the final step: connecting to the Lime SDR when launching a GNU Radio script!

To be exact, the error is along the lines of:
![image](https://github.com/user-attachments/assets/54700a6a-840c-46a7-87d5-023eb8559565)

This error tells us that the pass-through works (the USB device appears in WSL) but there's a permission error.

# Solution

A fix for this problem is presented below, by creating a new udev rule.

### 0 - Connect the device to WSL via `usbipd`

We will need some info of the device, and so we need it to be connected for the next step.
The method on how to do it is already listed [here](https://github.com/LELEC210X/LELEC210X/blob/main/INSTALL.md#only-for-wsl-users-connect-the-limesdr-to-the-wsl-via-usb-pass-through) in the [INSTALL.md](https://github.com/LELEC210X/LELEC210X/blob/main/INSTALL.md), but here's a refresher of the useful commands (in Windows' PowerShell):

```bat
usbipd list
usbipd bind --busid <busid>
usbipd attach --wsl --busid <busid>
usbipd detach --busid <busid>
```

### 1 - Retrieve the Lime SDR info
As the USB id will change every time we plug in the Lime SDR, we need an automated way to distinguish it from other devices.
This is done by retrieving its vendor id (VID) and product id (PID). To do so, you can use the following command (in WSL):
```bash
lsusb
```
which will return something along the lines of:
```bash
Bus 002 Device 002: ID AAAA:BBBB Future Technology Devices International, Ltd LimeSDR Mini
```
where `AAAA` is the VID and `BBBB` the PID. Note them down.

### 2 - Create the udev rule
To create the udev rule, you will need to write a `.rules` file at `/etc/udev/rules.d/`. This requires the use of elevated privileges.
I'll use `nano` in my case, to create the rule file `LELEC210X-limesdr-usbipd.rules`:
```bash
sudo nano /etc/udev/rules.d/LELEC210X-limesdr-usbipd.rules
```
In there, we can write the following (remember to replace `AAAA` and `BBBB` by their actual values):
```bash
SUBSYSTEM=="usb", ATTR{idVendor}=="AAAA", ATTR{idProduct}=="BBBB", MODE="0666"
```
and save it. **Don't forget** to use the correct permissions (`sudo`) or the contents of the file won't be saved !

### 3 - Reload the udev rules
As a final step, you will need to reload the udev rules.
This can be done by running either of those two sets of commands (choose the one that works with you).
```bash
sudo udevadm control --reload-rules
sudo udevadm trigger
``` 
or
```bash
sudo service udev restart
```
**Note**: simply restarting WSL might not be sufficient, reloading is important !

### 4 - Testing the new rule
We can check if this was done correctly by inspecting the file permissions of the Lime SDR once attached to WSL, for example using:
```bash
ll /dev/bus/usb/002
```
The Lime SDR should appear under the name `xxx` (e.g. `002`) with the correct permissions:
```bash
crw-rw-rw- 1 root root 189, 131 Jan 01 00:01 xxx
```
instead of the regular:
```bash
crw------- 1 root root 189, 131 Jan 01 00:01 xxx
```
**Note**: the name of the folder might not be the same.

---

This is all that I have to say, and I hope that it will help people sharing the same problem.

Have a nice day !

**Comment by jeertmans** [comment, marked as answer]
Hi @Reshims, thanks for providing a very well documented solution! I edited the post a bit, and added a mention to it in the Wiki!
